client.go

```diff
@@ -0,0 +1,354 @@
+	defaultResponseChunkingSupported = true
+)
+
+// NewADSClient creates a new [*ADSClient] with the given options. To stop the client, close the
+// backing [grpc.ClientConn].
+func NewADSClient(conn *grpc.ClientConn, node *ads.Node, opts ...ADSClientOption) *ADSClient {
+	c := &ADSClient{
+		conn:            conn,
+		node:            node,
+		newSubscription: make(chan struct{}, 1),
+		handlers:        make(map[string]internal.RawResourceHandler),
+		options: options{
+			initialReconnectBackoff:   defaultInitialReconnectBackoff,
+			maxReconnectBackoff:       defaultMaxReconnectBackoff,
+			responseChunkingSupported: defaultResponseChunkingSupported,
+		},
+	}
+
+	for _, opt := range opts {
+		opt(&c.options)
+	}
+
+	go c.loop()
+
+	return c
+}
+
+type options struct {
+	initialReconnectBackoff   time.Duration
+	maxReconnectBackoff       time.Duration
+	responseChunkingSupported bool
+}
+
+// WithReconnectBackoff provides backoff configuration when reconnecting to the xDS backend after a
+// connection failure. The default settings are 100ms and 2m for the initial and max backoff
+// respectively.
+func WithReconnectBackoff(initialBackoff, maxBackoff time.Duration) ADSClientOption {
+	return func(o *options) {
+		o.initialReconnectBackoff = initialBackoff
+		o.maxReconnectBackoff = maxBackoff
+	}
+}
+
+// WithResponseChunkingSupported changes whether response chunking should be supported (see
+// [ads.ParseRemainingChunksFromNonce] for additional details). This feature is only provided by the
+// [ADSServer] implemented in this package. This enabled by default.
+func WithResponseChunkingSupported(supported bool) ADSClientOption {
+	return func(o *options) {
+		o.responseChunkingSupported = supported
+	}
+}
+
+// An ADSClient is a client that implements the xDS protocol, and can therefore be used to talk to
+// any xDS backend. Use the [Watch], [WatchGlob] and [WatchWildcard] to subscribe to resources.
+type ADSClient struct {
+	options
+	node *ads.Node
+	conn *grpc.ClientConn
+
+	newSubscription chan struct{}
+
+	lock     sync.Mutex
+	handlers map[string]internal.RawResourceHandler
+}
+
+// A Watcher is used to receive updates from the xDS backend using an [ADSClient]. It is passed into
+// the various [Watch] methods in this package. Note that it is imperative that implementations be
+// hashable as it will be stored as the key to a map (unhashable types include slices and functions).
+type Watcher[T proto.Message] interface {
+	// Notify is invoked whenever a response is processed. The given sequence will iterate over all the
+	// resources in the response, with a nil resource indicating a deletion. Implementations should
+	// return an error if any resource is invalid, and this error will be propagated as a NACK to the xDS
+	// backend.
+	Notify(resources iter.Seq2[string, *ads.Resource[T]]) error
+}
+
+// Watch functions much like [Cache.Subscribe]. The given [Watcher] will be notified whenever a
```

Comment:
> Watch functions much like [Cache.Subscribe].

IMHO, I think this is misleading as Watch is to be used by client users, whereas the Cache is completely internal in the server.

Reply:
Sure, I can make this less dependent on the cache documentation